In [1]:
import numpy as np

In [3]:
np.__version__

'1.26.4'

Let's start by instantiating the same array from the previous module, which represents the performance of 5 athletes in 4 laps of a race track, on 3 different days.

In [4]:
rng = np.random.default_rng(seed=1)

In [5]:
array = rng.normal(loc=50, scale=3, size=60).reshape(3, 5, 4).round(3)

In [6]:
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

In [7]:
array.shape

(3, 5, 4)

👆 Notice from the `shape` that there are 3 days, 5 athletes and 4 laps.

### Conversion methods

#### `item`

Use the `item` method to recover the second athlete's performance on the third day, in the first lap.

In [8]:
array.item((2, 1, 0))

49.705

#### `itemset`

Suppose this value was recorded incorrectly, and the correct value is 49.709. Make the correction using `itemset`, and display the corrected array.

In [9]:
array.itemset((2, 1, 0), 49.709)

In [10]:
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.709, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

Use `item` again to make sure the value has changed.

In [11]:
array.item((2, 1, 0))

49.709

#### `view`

Generate a view of this array, assigning the result to a new array `array_view`.

In [12]:
array_view = array.view()

Use `itemset` to change the previously corrected value (index (2, 1, 0)) to its original value (49,705), only in `array_view`.

In [13]:
array_view.itemset((2, 1, 0), 49.705)

Now use `item` to check the value at index (2, 1, 0) in `array` and `array_view`.

In [14]:
array.item((2, 1, 0))

49.705

In [ ]:
array_view.item((2, 1, 0))

49.705

💡 Note that although you changed the value in `array_view`, `array` was also modified. This happens because, as `array_view` is a **view** of `array`, which means that its data is shared.

#### `copy`

Now repeat the previous exercise, but create a copy of `array`, stored in `array_copy`.

In [15]:
array_copy = array.copy()

Change the value in (2, 1, 0) back to 49.709, only in the copy.

In [16]:
array_copy.itemset((2, 1, 0), 49.709)

Check with `view` that this value is now different in `array` and `array_copy`.

In [17]:
array.item((2, 1, 0))

49.705

In [18]:
array_copy.item((2, 1, 0))

49.709

💡 A **copy** of an array holds data independent of the original array.

### Methods to manipulate `shape`

#### `reshape`

Suppose you need to record the athletes' performance results in a spreadsheet, which stores data in only 2 dimensions. The guidance you received is that each row of the spreadsheet must correspond to a day, totaling 3 rows, and the columns will contain the 4 laps of the first athlete, then the 4 laps of the second athlete, and so on, totaling 5 athletes x 4 loops = 20 columns.

Display a version of the array in this configuration.

In [19]:
array.reshape(3, 20)

array([[51.037, 52.465, 50.991, 46.091, 52.716, 51.339, 48.389, 51.743,
        51.094, 50.882, 50.085, 51.64 , 47.791, 49.511, 48.554, 51.797,
        50.119, 49.123, 47.654, 49.228],
       [50.024, 49.173, 53.882, 53.02 , 41.867, 44.333, 49.476, 48.733,
        50.641, 50.652, 56.354, 46.664, 48.867, 56.128, 51.94 , 51.989,
        48.458, 45.056, 50.502, 50.327],
       [46.318, 47.95 , 49.784, 47.166, 49.705, 50.286, 50.107, 48.481,
        51.781, 52.674, 50.963, 47.545, 52.195, 48.496, 52.637, 46.785,
        52.743, 49.94 , 46.254, 49.058]])

#### `transpose`

Remember that the dimensions of the array represent the days, athletes and laps, in that order.

Use `transpose` to represent the array with the athletes' data, then the laps, then the days. Store the result in a new `array_transpose`, display the array and then its `shape`.

In [21]:
array.shape

(3, 5, 4)

In [20]:
array_transpose = array.transpose((1, 2, 0))

In [ ]:
array_transpose

array([[[51.037, 50.024, 46.318],
        [52.465, 49.173, 47.95 ],
        [50.991, 53.882, 49.784],
        [46.091, 53.02 , 47.166]],

       [[52.716, 41.867, 49.705],
        [51.339, 44.333, 50.286],
        [48.389, 49.476, 50.107],
        [51.743, 48.733, 48.481]],

       [[51.094, 50.641, 51.781],
        [50.882, 50.652, 52.674],
        [50.085, 56.354, 50.963],
        [51.64 , 46.664, 47.545]],

       [[47.791, 48.867, 52.195],
        [49.511, 56.128, 48.496],
        [48.554, 51.94 , 52.637],
        [51.797, 51.989, 46.785]],

       [[50.119, 48.458, 52.743],
        [49.123, 45.056, 49.94 ],
        [47.654, 50.502, 46.254],
        [49.228, 50.327, 49.058]]])

In [22]:
array_transpose.shape

(5, 4, 3)

👆 5 athletes, 4 laps, 3 days.

💡 That is, the first row of the visualization:
> ```
> [51.037, 50.024, 46.318]
> ```

represents the first athlete, the first lap, and the 3 days.

#### `swapaxes`

From `array_transpose`, now use `swapaxes` to change the axes to represent athletes, then days, then laps. Store the result in `array_swapaxes`, display the array and its shape.

In [23]:
array_swapaxes = array_transpose.swapaxes(2, 1)

In [24]:
array_swapaxes

array([[[51.037, 52.465, 50.991, 46.091],
        [50.024, 49.173, 53.882, 53.02 ],
        [46.318, 47.95 , 49.784, 47.166]],

       [[52.716, 51.339, 48.389, 51.743],
        [41.867, 44.333, 49.476, 48.733],
        [49.705, 50.286, 50.107, 48.481]],

       [[51.094, 50.882, 50.085, 51.64 ],
        [50.641, 50.652, 56.354, 46.664],
        [51.781, 52.674, 50.963, 47.545]],

       [[47.791, 49.511, 48.554, 51.797],
        [48.867, 56.128, 51.94 , 51.989],
        [52.195, 48.496, 52.637, 46.785]],

       [[50.119, 49.123, 47.654, 49.228],
        [48.458, 45.056, 50.502, 50.327],
        [52.743, 49.94 , 46.254, 49.058]]])

In [25]:
array_swapaxes.shape

(5, 3, 4)

👆 5 athletes, 3 days, 4 laps.

💡 That is, the first row of the visualization:
> ```
> [51.037, 52.465, 50.991, 46.091]
> ```

represents the first athlete, the first day, the 4 laps.

#### `flatten` / `ravel`

Now suppose you have to record the results in a system that has been programmed to receive data linearly, one value at a time.

Linearize the array using `flatten` and `ravel`.

In [26]:
array.flatten()

array([51.037, 52.465, 50.991, 46.091, 52.716, 51.339, 48.389, 51.743,
       51.094, 50.882, 50.085, 51.64 , 47.791, 49.511, 48.554, 51.797,
       50.119, 49.123, 47.654, 49.228, 50.024, 49.173, 53.882, 53.02 ,
       41.867, 44.333, 49.476, 48.733, 50.641, 50.652, 56.354, 46.664,
       48.867, 56.128, 51.94 , 51.989, 48.458, 45.056, 50.502, 50.327,
       46.318, 47.95 , 49.784, 47.166, 49.705, 50.286, 50.107, 48.481,
       51.781, 52.674, 50.963, 47.545, 52.195, 48.496, 52.637, 46.785,
       52.743, 49.94 , 46.254, 49.058])

In [27]:
array.ravel()

array([51.037, 52.465, 50.991, 46.091, 52.716, 51.339, 48.389, 51.743,
       51.094, 50.882, 50.085, 51.64 , 47.791, 49.511, 48.554, 51.797,
       50.119, 49.123, 47.654, 49.228, 50.024, 49.173, 53.882, 53.02 ,
       41.867, 44.333, 49.476, 48.733, 50.641, 50.652, 56.354, 46.664,
       48.867, 56.128, 51.94 , 51.989, 48.458, 45.056, 50.502, 50.327,
       46.318, 47.95 , 49.784, 47.166, 49.705, 50.286, 50.107, 48.481,
       51.781, 52.674, 50.963, 47.545, 52.195, 48.496, 52.637, 46.785,
       52.743, 49.94 , 46.254, 49.058])

💡 Remember that `flatten` returns a **copy** of the array, while `ravel` returns a **view**.

#### `squeeze`

We can separate the first day's results into a new array `array_first_day`, with the following notation:

In [28]:
array_first_day = array[0:1] # instead of simply array[0]

In [29]:
array_first_day

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]]])

In [30]:
array_first_day.ndim, array_first_day.shape

(3, (1, 5, 4))

Note that the number of dimensions (`ndim`) of `array_first_day` remains equal to 3, that is, the same as `array`. With `shape`, we observe that the axis corresponding to days has a single element, precisely the first day selected. The slicing of a single value with the notation `[n:n+1]` returns an array that explicitly represents a single element, as further revealed by `shape`.

Use `squeeze` to simplify this array, removing the days axis, as it only contains a single day. Save the result in `array_squeeze`, display the array and then its `shape`.

In [31]:
array_squeeze = array_first_day.squeeze()

In [33]:
array_squeeze

array([[51.037, 52.465, 50.991, 46.091],
       [52.716, 51.339, 48.389, 51.743],
       [51.094, 50.882, 50.085, 51.64 ],
       [47.791, 49.511, 48.554, 51.797],
       [50.119, 49.123, 47.654, 49.228]])

👆 Notice that there is one less pair of brackets, which indicates a reduction of 1 axis.

In [34]:
array_squeeze.shape

(5, 4)

👆 The axis corresponding to the day, of size 1, has been removed.

### Methods to select and manipulate items

#### `take`

Now consider that the data in `array` refers to the same athlete. The first axis continues to represent 3 different days, the last represents the 4 laps of a race, but the second axis represents 5 complete races of the same athlete. In other words, `array` now contains consecutive data from the same athlete.

Suppose that, as soon as he finished the tests, the athlete complained of knee pain on laps 12, 25, 31 and 40. Using `take`, recover the athlete's performance on these laps.

In [35]:
indexesakeiindexes=[12, 25, 31, 40])indexesindexes

array([47.791, 44.333, 46.664, 46.318])

But the athlete was more specific and said that, on laps 12 and 31, the pain was in the right knee, and on laps 25 and 40, the pain was in the left knee. Using `take`, retrieve these laps again, but in order to obtain the results as an array where the first row refers to the right knee, and the second to the left knee.

In [36]:
array.take(indexes=[[12, 31],
                    [25, 40]])

array([[47.791, 46.664],
       [44.333, 46.318]])

#### `put`

Based on what the athlete reported, his coach decided to disregard the athlete's performance on the laps where he complained of pain. Use `put` to change these values ​​to the constant `np.nan`, which represents null values ​​in a NumPy array. Then display the modified array.

In [37]:
array.put(indexes=[12, 25, 31, 40], values = np.nan)

In [38]:
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867,    nan, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

#### `repeat`

Repeat the array values ​​so that they are duplicated (`repeat = 2`), along axis 0, 1 and 2, separately.

Then display the array to inspect where the repetition occurred. Also display the `shape` of the array to observe the change made.

In [39]:
array_repeat_axis_0 = array.repeat(repeats=2, axis=0)

In [40]:
array_repeat_axis_0

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867,    nan, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867,    nan, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 5

In [41]:
array_repeat_axis_0.shape

(6, 5, 4)

👆 Notice that the data corresponding to the days has been duplicated.

In [42]:
array_repeat_axis_1 = array.repeat(repeats=2, axis=1)

In [43]:
array_repeat_axis_1

array([[[51.037, 52.465, 50.991, 46.091],
        [51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [51.094, 50.882, 50.085, 51.64 ],
        [   nan, 49.511, 48.554, 51.797],
        [   nan, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [50.024, 49.173, 53.882, 53.02 ],
        [41.867,    nan, 49.476, 48.733],
        [41.867,    nan, 49.476, 48.733],
        [50.641, 50.652, 56.354,    nan],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 56.128, 51.94 , 51.989],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327],
        [48.458, 45.056, 50.502, 50.327]],

       [[   nan, 47.95 , 49.784, 47.166],
        [   nan, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [49.705, 50.286, 50.10

In [44]:
array_repeat_axis_1.shape

(3, 10, 4)

👆 Notice that the data corresponding to the tests has been duplicated.

In [45]:
array_repeat_axis_2 = array.repeat(repeats=2, axis=2)

In [46]:
array_repeat_axis_2

array([[[51.037, 51.037, 52.465, 52.465, 50.991, 50.991, 46.091, 46.091],
        [52.716, 52.716, 51.339, 51.339, 48.389, 48.389, 51.743, 51.743],
        [51.094, 51.094, 50.882, 50.882, 50.085, 50.085, 51.64 , 51.64 ],
        [   nan,    nan, 49.511, 49.511, 48.554, 48.554, 51.797, 51.797],
        [50.119, 50.119, 49.123, 49.123, 47.654, 47.654, 49.228, 49.228]],

       [[50.024, 50.024, 49.173, 49.173, 53.882, 53.882, 53.02 , 53.02 ],
        [41.867, 41.867,    nan,    nan, 49.476, 49.476, 48.733, 48.733],
        [50.641, 50.641, 50.652, 50.652, 56.354, 56.354,    nan,    nan],
        [48.867, 48.867, 56.128, 56.128, 51.94 , 51.94 , 51.989, 51.989],
        [48.458, 48.458, 45.056, 45.056, 50.502, 50.502, 50.327, 50.327]],

       [[   nan,    nan, 47.95 , 47.95 , 49.784, 49.784, 47.166, 47.166],
        [49.705, 49.705, 50.286, 50.286, 50.107, 50.107, 48.481, 48.481],
        [51.781, 51.781, 52.674, 52.674, 50.963, 50.963, 47.545, 47.545],
        [52.195, 52.195, 48.496, 4

In [48]:
array_repeat_axis_2.shape

(3, 5, 8)

👆 Notice that the data corresponding to the laps has been duplicated.

#### `sort`

First, create a copy of `array`, since `sort` changes the array *inplace*.

Then, use `sort` to sort the array along axis 0. Display the sorted array to see what happened.

In [49]:
array_copy = array.copy()

In [50]:
array_copy.sort(axis=0)

In [51]:
array_copy

array([[[50.024, 47.95 , 49.784, 46.091],
        [41.867, 50.286, 48.389, 48.481],
        [50.641, 50.652, 50.085, 47.545],
        [48.867, 48.496, 48.554, 46.785],
        [48.458, 45.056, 46.254, 49.058]],

       [[51.037, 49.173, 50.991, 47.166],
        [49.705, 51.339, 49.476, 48.733],
        [51.094, 50.882, 50.963, 51.64 ],
        [52.195, 49.511, 51.94 , 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[   nan, 52.465, 53.882, 53.02 ],
        [52.716,    nan, 50.107, 51.743],
        [51.781, 52.674, 56.354,    nan],
        [   nan, 56.128, 52.637, 51.989],
        [52.743, 49.94 , 50.502, 50.327]]])

👆 Note that, over the days, the values ​​are ordered in ascending order:

`50.024 -> 51.037 -> nan`

`47.95 -> 49.173 -> 52.465`...

Also note that NumPy places `nan` at the **end** of the order; all `nan` values ​​are on the third element of axis 0.

Now repeat the procedure, copying the array and then sorting along axis 1.

In [52]:
array_copy = array.copy()

In [53]:
array_copy.sort(axis=1)

In [54]:
array_copy

array([[[50.119, 49.123, 47.654, 46.091],
        [51.037, 49.511, 48.389, 49.228],
        [51.094, 50.882, 48.554, 51.64 ],
        [52.716, 51.339, 50.085, 51.743],
        [   nan, 52.465, 50.991, 51.797]],

       [[41.867, 45.056, 49.476, 48.733],
        [48.458, 49.173, 50.502, 50.327],
        [48.867, 50.652, 51.94 , 51.989],
        [50.024, 56.128, 53.882, 53.02 ],
        [50.641,    nan, 56.354,    nan]],

       [[49.705, 47.95 , 46.254, 46.785],
        [51.781, 48.496, 49.784, 47.166],
        [52.195, 49.94 , 50.107, 47.545],
        [52.743, 50.286, 50.963, 48.481],
        [   nan, 52.674, 52.637, 49.058]]])

👆 Now, the ordering occurs along the tests (which in the visualization correspond to the columns):

`50,119 -> 51,037 -> 51,094 -> 52,716 -> nan`...

This time, the `nan` values ​​are in the last element of axis 1 (last row of each array representing days).

Repeat once more, copying the array and then sorting along axis 2.

In [55]:
array_copy = array.copy()

In [56]:
array_copy.sort(axis=2)

In [57]:
array_copy

array([[[46.091, 50.991, 51.037, 52.465],
        [48.389, 51.339, 51.743, 52.716],
        [50.085, 50.882, 51.094, 51.64 ],
        [48.554, 49.511, 51.797,    nan],
        [47.654, 49.123, 49.228, 50.119]],

       [[49.173, 50.024, 53.02 , 53.882],
        [41.867, 48.733, 49.476,    nan],
        [50.641, 50.652, 56.354,    nan],
        [48.867, 51.94 , 51.989, 56.128],
        [45.056, 48.458, 50.327, 50.502]],

       [[47.166, 47.95 , 49.784,    nan],
        [48.481, 49.705, 50.107, 50.286],
        [47.545, 50.963, 51.781, 52.674],
        [46.785, 48.496, 52.195, 52.637],
        [46.254, 49.058, 49.94 , 52.743]]])

👆 The ordering occurs along the laps (which in the visualization correspond to the rows):

`46,091 -> 50,991 -> 51,037 -> 52,465`...

This time, the `nan` values ​​are in the last element of axis 2 (last column of each array representing days).

#### `argsort`

Repeat the `sort` exercise, but now, instead of sorting the values *inplace*, retrieve and display the indexes necessary to sort the array. This is done with the `argsort` method.

This time, since there is no modification of the array, you don't need the copy.

In [58]:
indexes_axis_0 = array.argsort(axis=0)

In [59]:
indexes_axis_0

array([[[1, 2, 2, 0],
        [1, 2, 0, 2],
        [1, 1, 0, 2],
        [1, 2, 0, 2],
        [1, 1, 2, 2]],

       [[0, 1, 0, 2],
        [2, 0, 1, 1],
        [0, 0, 2, 0],
        [2, 0, 1, 0],
        [0, 0, 0, 0]],

       [[2, 0, 1, 1],
        [0, 1, 2, 0],
        [2, 2, 1, 1],
        [0, 1, 2, 1],
        [2, 2, 1, 1]]])

👆 Note that the maximum value in this array representing the indexes is 2, because axis 0 only contains 3 elements (3 days).

In [60]:
indexes_axis_1 = array.argsort(axis=1)

In [61]:
indexes_axis_1

array([[[4, 4, 4, 0],
        [0, 3, 1, 4],
        [2, 2, 3, 2],
        [1, 1, 2, 1],
        [3, 0, 0, 3]],

       [[1, 4, 1, 1],
        [4, 0, 4, 4],
        [3, 2, 3, 3],
        [0, 3, 0, 0],
        [2, 1, 2, 2]],

       [[1, 0, 4, 3],
        [2, 3, 0, 0],
        [3, 4, 1, 2],
        [4, 1, 2, 1],
        [0, 2, 3, 4]]])

👆 Now the maximum value is 4, as axis 1 has 5 elements (5 athletes).

In [62]:
indexes_axis_2 = array.argsort(axis=2)

In [63]:
indexes_axis_2

array([[[3, 2, 0, 1],
        [2, 1, 3, 0],
        [2, 1, 0, 3],
        [2, 1, 3, 0],
        [2, 1, 3, 0]],

       [[1, 0, 3, 2],
        [0, 3, 2, 1],
        [0, 1, 2, 3],
        [0, 2, 3, 1],
        [1, 0, 3, 2]],

       [[3, 1, 2, 0],
        [3, 0, 2, 1],
        [3, 2, 0, 1],
        [3, 1, 0, 2],
        [2, 3, 1, 0]]])

👆 The maximum value is 3, as axis 2 has 4 elements (4 turns).

#### `searchsorted`

This method only works for 1D arrays that are already sorted.

So, first linearize `array` with `flatten`, then sort it with `sort` (remember that the `sort` method alters the array *inplace*, i.e. it doesn't return anything, so you don't need to store the result in a new variable), and use indexing to retrieve the first 15 elements.

In [64]:
array_flatten = array.flatten()

In [65]:
array_flatten

array([51.037, 52.465, 50.991, 46.091, 52.716, 51.339, 48.389, 51.743,
       51.094, 50.882, 50.085, 51.64 ,    nan, 49.511, 48.554, 51.797,
       50.119, 49.123, 47.654, 49.228, 50.024, 49.173, 53.882, 53.02 ,
       41.867,    nan, 49.476, 48.733, 50.641, 50.652, 56.354,    nan,
       48.867, 56.128, 51.94 , 51.989, 48.458, 45.056, 50.502, 50.327,
          nan, 47.95 , 49.784, 47.166, 49.705, 50.286, 50.107, 48.481,
       51.781, 52.674, 50.963, 47.545, 52.195, 48.496, 52.637, 46.785,
       52.743, 49.94 , 46.254, 49.058])

In [66]:
array_flatten.sort()

In [67]:
array_flatten

array([41.867, 45.056, 46.091, 46.254, 46.785, 47.166, 47.545, 47.654,
       47.95 , 48.389, 48.458, 48.481, 48.496, 48.554, 48.733, 48.867,
       49.058, 49.123, 49.173, 49.228, 49.476, 49.511, 49.705, 49.784,
       49.94 , 50.024, 50.085, 50.107, 50.119, 50.286, 50.327, 50.502,
       50.641, 50.652, 50.882, 50.963, 50.991, 51.037, 51.094, 51.339,
       51.64 , 51.743, 51.781, 51.797, 51.94 , 51.989, 52.195, 52.465,
       52.637, 52.674, 52.716, 52.743, 53.02 , 53.882, 56.128, 56.354,
          nan,    nan,    nan,    nan])

In [68]:
array_filtered = array_flatten[:15]

In [69]:
array_filtered

array([41.867, 45.056, 46.091, 46.254, 46.785, 47.166, 47.545, 47.654,
       47.95 , 48.389, 48.458, 48.481, 48.496, 48.554, 48.733])

Using `searchsorted`, find out in which positions the values ​​40.50, 42.357, 47.025 and 48.952 should be inserted to maintain the sorting.

In [70]:
indexes = array_filtered.searchsorted([40.50, 42.357, 47.025, 48.952])

In [71]:
indexes

array([ 0,  1,  5, 15])

👆 Inspect the array and the indexes found to confirm the answer.

#### `diagonal`

Understanding diagonals is made easier by working with 2D arrays.

Therefore, first obtain with indexing a view of the array corresponding to the first day.

In [72]:
array_2D = array[0]

In [ ]:
array_2D

array([[51.037, 52.465, 50.991, 46.091],
       [52.716, 51.339, 48.389, 51.743],
       [51.094, 50.882, 50.085, 51.64 ],
       [   nan, 49.511, 48.554, 51.797],
       [50.119, 49.123, 47.654, 49.228]])

In [73]:
array_2D.shape

(5, 4)

Now retrieve and display the diagonal of this array.

In [74]:
array_2D.diagonal()

array([51.037, 51.339, 50.085, 51.797])

👆 It is not difficult to visually find this diagonal in `array_2D`.

Note that, as `array_2D` is **rectangular** (as demonstrated by its `shape`), the diagonal does not contain any value from the last row, as theh array has more rows than columns.